File: sdks/node/test/core/contractFloor.test.ts
// The storage carries the lowest contract level a process may speak.
//
// A build below that floor must refuse to open rather than join a deployment
// and misread rows its contract never described.

import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, describe, expect, it } from "vitest";
import { Queue } from "../../src/index";

const CONTRACT_FLOOR_SETTING = "contract:min_sdk";

let tempDir: string | undefined;

afterEach(() => {
  if (tempDir) {
    // Best effort: the queue's native SQLite handle is released on GC, which
    // JS cannot force, and Windows refuses to unlink a file that still has one.
    try {
      rmSync(tempDir, { recursive: true, force: true, maxRetries: 5, retryDelay: 50 });
    } catch {
      // the OS reclaims it
    }
    tempDir = undefined;
  }
});

function openQueue(): { queue: Queue; dbPath: string } {
  tempDir ??= mkdtempSync(join(tmpdir(), "flexiq-contract-"));
  const dbPath = join(tempDir, "q.db");
  return { queue: new Queue({ dbPath }), dbPath };
}

describe("contract floor", () => {
  it("records the level that created the database", () => {
    // Creating the schema is the one moment the floor moves on its own: a
    // database that did not exist has no older process to lock out, and this
    // is what lets durable steps work without an operator finding the dial.
    const { queue } = openQueue();
    expect(queue.getSetting(CONTRACT_FLOOR_SETTING)).not.toBeNull();
    expect(queue.minContract()).toBeGreaterThanOrEqual(1);
  });

  it("leaves an unraised floor unwritten", () => {
    // A database an earlier release created carries no row, and opening never
    // adds one — the dial stays the operator's on an existing deployment.
    const { queue } = openQueue();
    queue.deleteSetting(CONTRACT_FLOOR_SETTING);

    expect(queue.getSetting(CONTRACT_FLOOR_SETTING)).toBeNull();
    expect(queue.minContract()).toBeGreaterThanOrEqual(1);
  });

  it("still opens storage whose floor is exactly this build", () => {
    const { queue, dbPath } = openQueue();
    queue.setMinContract(queue.minContract());

    expect(new Queue({ dbPath }).minContract()).toBe(queue.minContract());
  });

  it("refuses to open storage that requires a newer build", () => {
    const { queue, dbPath } = openQueue();
    const unreachable = queue.minContract() + 1;
    // Written through the raw setting: `setMinContract` rejects a level this
    // build cannot speak, which is what the next test exercises.
    queue.setSetting(CONTRACT_FLOOR_SETTING, String(unreachable));

    expect(() => new Queue({ dbPath })).toThrow(new RegExp(`contract ${unreachable}`));
  });

  it("rejects a floor this build cannot speak", () => {
    const { queue } = openQueue();
    const before = queue.minContract();

    expect(() => queue.setMinContract(before + 1)).toThrow(/lock it out/);
    expect(queue.minContract()).toBe(before);
  });
});

File: sdks/java/src/test/java/org/byteveda/flexiq/ContractFloorTest.java
package org.byteveda.flexiq;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.nio.file.Path;
import java.util.Map;
import org.byteveda.flexiq.internal.JniQueueBackend;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

/**
 * The storage carries the lowest contract level a process may speak.
 *
 * <p>A build below that floor must refuse to open rather than join a deployment
 * and misread rows its contract never described.
 */
class ContractFloorTest {

    private static final String CONTRACT_FLOOR_SETTING = "contract:min_sdk";

    private static String optionsFor(Path dir) throws Exception {
        return new ObjectMapper()
                .writeValueAsString(
                        Map.of("backend", "sqlite", "dsn", dir.resolve("t.db").toString()));
    }

    @Test
    void leavesAnUnraisedFloorUnwritten(@TempDir Path dir) throws Exception {
        // Opening never writes, so a deployment that leaves the dial alone
        // carries no row for it.
        try (FlexiQ queue = FlexiQ.builder().open(JniQueueBackend.open(optionsFor(dir)))) {
            assertTrue(queue.getSetting(CONTRACT_FLOOR_SETTING).isEmpty());
            assertTrue(queue.minContract() >= 1);
        }
    }

    @Test
    void stillOpensStorageWhoseFloorIsExactlyThisBuild(@TempDir Path dir) throws Exception {
        String options = optionsFor(dir);
        int level;
        try (FlexiQ queue = FlexiQ.builder().open(JniQueueBackend.open(options))) {
            level = queue.minContract();
            queue.setMinContract(level);
        }

        try (FlexiQ reopened = FlexiQ.builder().open(JniQueueBackend.open(options))) {
            assertEquals(level, reopened.minContract());
        }
    }

    @Test
    void refusesToOpenStorageThatRequiresANewerBuild(@TempDir Path dir) throws Exception {
        String options = optionsFor(dir);
        int unreachable;
        try (FlexiQ queue = FlexiQ.builder().open(JniQueueBackend.open(options))) {
            unreachable = queue.minContract() + 1;
            // Written through the raw setting: setMinContract rejects a level
            // this build cannot speak, which the next test exercises.
            queue.setSetting(CONTRACT_FLOOR_SETTING, String.valueOf(unreachable));
        }

        Exception failure = assertThrows(Exception.class, () -> JniQueueBackend.open(options));
        assertTrue(
                failure.getMessage().contains(String.valueOf(unreachable)),
                "the error must name the required level: " + failure.getMessage());
    }

    @Test
    void rejectsAFloorThisBuildCannotSpeak(@TempDir Path dir) throws Exception {
        try (FlexiQ queue = FlexiQ.builder().open(JniQueueBackend.open(optionsFor(dir)))) {
            int before = queue.minContract();

            Exception failure = assertThrows(Exception.class, () -> queue.setMinContract(before + 1));
            assertTrue(failure.getMessage().contains("lock it out"), "unexpected message: " + failure.getMessage());
            assertEquals(before, queue.minContract());
        }
    }
}
